lua/pl/tablex.lua

```diff
@@ -95,37 +95,50 @@ function tablex.copy (t)
     return res
 end
 
---- make a deep copy of a table, recursively copying all the keys and fields.
--- This will also set the copied table's metatable to that of the original.
--- @within Copying
--- @tab t A table
--- @return new table
-function tablex.deepcopy(t)
+local function cycle_aware_copy(t, res, cache)
     if type(t) ~= 'table' then return t end
     assert_arg_iterable(1,t)
     local mt = getmetatable(t)
-    local res = {}
     for k,v in pairs(t) do
+        local kk, vv = k, v
+        if type(k) == 'table' then
+            if cache[k] then
+                kk = cache[k]
+            else
+                kk = {}
+                cache[k] = kk
+                cycle_aware_copy(k,kk,cache)
+            end
+        end
         if type(v) == 'table' then
-            v = tablex.deepcopy(v)
+            if cache[v] then
+                vv = cache[v]
+            else
+                vv = {}
+                cache[v] = vv
+                cycle_aware_copy(v,vv,cache)
```

Comment:
I don't like code duplication here. What if `cycle_aware_copy` handled cache logic for its argument `t` instead? Additionally, it could create result table itself. Then the inner `pairs` could be simply

```lua
k = cycle_aware_copy(k, cache)
v = cycle_aware_copy(v, cache)
res[k] = v
```

Reply:
Ah, if key are not recursively then it does not matter too much I guess.

Reply:
@hishamhm actually, I think it's still a good idea to move the caching logic to the top level, otherwise the root table is not immediately cached.

Reply:
@hishamhm I think @mpeterv is right, the current implementation doesn't cache the top-level table

Reply:
@mpeterv @Tieske should be fixed, thanks for the nudge!